File: app/src/main/java/uk/co/mishurov/termik/MainActivity.java
package uk.co.mishurov.termik;


import java.io.File;
import java.util.List;

import android.Manifest;
import android.content.Intent;
import android.os.Bundle;
import android.os.Environment;
import android.os.storage.OnObbStateChangeListener;
import android.os.storage.StorageManager;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.SystemClock;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.SurfaceView;
import android.view.WindowManager;
import android.view.Display;
import android.view.Surface;
import android.view.OrientationEventListener;
import android.hardware.SensorManager;
import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.util.DisplayMetrics;

import com.google.android.vending.expansion.downloader.Helpers;

import org.opencv.android.BaseLoaderCallback;
import org.opencv.android.CameraBridgeViewBase;
import org.opencv.android.LoaderCallbackInterface;
import org.opencv.android.OpenCVLoader;
import org.opencv.core.Mat;
import org.opencv.core.Size;
import org.opencv.core.Rect;
import org.opencv.core.Core;
import org.opencv.core.CvType;
import org.opencv.imgproc.Imgproc;
import org.opencv.android.Utils;

import org.tensorflow.demo.Classifier;
import org.tensorflow.demo.TensorFlowImageClassifier;

import android.util.Log;


public class MainActivity extends AppCompatActivity
                          implements CameraBridgeViewBase.CvCameraViewListener2 {

    private static final String TAG = "Termik";

    // Orientation
    private OrientationEventListener mOrientationListener;
    private int mOrientation = 0;
    private int mScreenRotation = 0;

    // Assets
    private final static String EXP_PATH = "/Android/obb/";
    private final static int VERSION_CODE = 1;
    private static String mObbPath;
    private String mStatus;
    private String mMountedPath;
    private StorageManager mSM;

    // Classification
    private static final String VISUAL = "Visual:\n";
    private static final String MODEL_FILENAME = "/tensorflow_inception_graph.pb";
    private static final String LABEL_FILENAME = "/imagenet_comp_graph_label_strings.txt";
    private static final int INPUT_SIZE = 224;
    private static final int IMAGE_MEAN = 117;
    private static final float IMAGE_STD = 1;
    private static final String INPUT_NAME = "input";
    private static final String OUTPUT_NAME = "output";
    private Classifier mClassifier = null;
    private Bitmap mInputBitmap = null;
    private Mat mRecentFrame = null;
    private Handler handler;
    private HandlerThread handlerThread;
    private int mPreviewWidth = 0;
    private int mPreviewHeight = 0;
    private boolean mProcessing = false;
    private JavaCameraView mCameraView;
    private ResultsView mVisuals;

    private static class XAPKFile {
        public final boolean mIsMain;
        public final int mFileVersion;
        public final long mFileSize;

        XAPKFile(boolean isMain, int fileVersion, long fileSize) {
            mIsMain = isMain;
            mFileVersion = fileVersion;
            mFileSize = fileSize;
        }
    }

    private static final XAPKFile[] xAPKS = {
        new XAPKFile(
            true, // true signifies a main file
            1, // the version of the APK that the file was uploaded against
            54112309L // the length of the file in bytes
        )
    };

    boolean expansionFilesDelivered() {
        for (XAPKFile xf : xAPKS) {
            String fileName = Helpers.getExpansionAPKFileName(
                this, xf.mIsMain, xf.mFileVersion
            );
            if (!Helpers.doesFileExist(this, fileName, xf.mFileSize, false)) {
                Log.e(
                    TAG,
                    "ExpansionAPKFile doesn't exist or has a wrong size (" + fileName + ")."
                );
                return false;
            }
        }
        return true;
    }

    private BaseLoaderCallback mBaseLoaderCallback = new BaseLoaderCallback(this) {
        @Override
        public void onManagerConnected(int status) {
            switch (status) {
                case LoaderCallbackInterface.SUCCESS: {
                    Log.i(TAG, "OpenCV loaded successfully");
                    // Load ndk built module, as specified in moduleName in build.gradle
                    // after opencv initialization
                    System.loadLibrary("processing");
                    mCameraView.enableView();
                }
                break;
                default: {
                    super.onManagerConnected(status);
                }
            }
        }
    };

    void setUpClassifier() {
        String modelFile = mMountedPath + MODEL_FILENAME;
        String labelFile = mMountedPath + LABEL_FILENAME;
        mClassifier = TensorFlowImageClassifier.create(
            getAssets(),
            modelFile,
            labelFile,
            INPUT_SIZE,
            IMAGE_MEAN,
            IMAGE_STD,
            INPUT_NAME,
            OUTPUT_NAME
        );
    }

    OnObbStateChangeListener mEventListener = new OnObbStateChangeListener() {
        @Override
        public void onObbStateChange(String path, int state) {
            mStatus = String.valueOf(state);
            if (state == OnObbStateChangeListener.MOUNTED) {
                mMountedPath = mSM.getMountedObbPath(mObbPath);
                setdir(mMountedPath);
                setUpClassifier();
                Log.d(TAG, "Mounted, path: " + mMountedPath);
            } else {
                mMountedPath = "";
                Log.d(TAG, "Mount failed, state: " + state);
            }
        }
    };

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().addFlags(
            WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
        );
        setContentView(R.layout.activity_main);

        // Permissions for Android 6+
        ActivityCompat.requestPermissions(
            MainActivity.this,
            new String[]{
                Manifest.permission.CAMERA,
                Manifest.permission.READ_EXTERNAL_STORAGE},
            1
        );

        // Check if expansion files are available before going any further
        if (!expansionFilesDelivered()) {
            Intent intent = new Intent(MainActivity.this, ProgressActivity.class);
            startActivity(intent);
        }

        ObbState state = (ObbState) getLastNonConfigurationInstance();
        if (state != null) {
            mSM = state.storageManager;
            mStatus = state.status.toString();
            mMountedPath = state.path.toString();
        } else {
            // Get an instance of the StorageManager
            mSM = (StorageManager) getApplicationContext().getSystemService(STORAGE_SERVICE);
        }

        // Get the path to the obb file
        File root = Environment.getExternalStorageDirectory();
        String path = root.toString() + EXP_PATH + getPackageName();
        String filename = "main."  + VERSION_CODE + "." + getPackageName() + ".obb";
        mObbPath = path + "/" + filename;

        if (mSM.mountObb(mObbPath, null, mEventListener)) {
            Log.d(TAG, "Starting mounting");
        } else {
            Log.d(TAG, "Start of mounting failed");
        }

        mCameraView = (JavaCameraView) findViewById(R.id.main_surface);
        mCameraView.setVisibility(SurfaceView.VISIBLE);
        mCameraView.setCvCameraViewListener(this);

        Display display = getWindowManager().getDefaultDisplay();
        int rotation = display.getRotation();
        switch (rotation) {
            case Surface.ROTATION_90:
                mScreenRotation = 90;
                break;
            case Surface.ROTATION_180:
                mScreenRotation = 180;
                break;
            case Surface.ROTATION_270:
                mScreenRotation = -90;
                break;
            default:
                mScreenRotation = 0;
                break;
        }

        // Listen orientation
        mVisuals = (ResultsView) findViewById(R.id.linear);
        mVisuals.setResults(VISUAL + "Calculating...");
        mOrientationListener = new OrientationEventListener(this,
            SensorManager.SENSOR_DELAY_NORMAL) {
                @Override
                public void onOrientationChanged(int orientation) {
                    orientation += mScreenRotation;
                    MainActivity.this.mOrientation = orientation;
                    MainActivity.this.mVisuals.adjust(orientation);
                }
            };

        if (mOrientationListener.canDetectOrientation()) {
            Log.v(TAG, "Can detect orientation");
            mOrientationListener.enable();
        } else {
            Log.v(TAG, "Cannot detect orientation");
            mOrientationListener.disable();
        }

    }

    @Override
    public synchronized void onPause() {
        handlerThread.quitSafely();
        try {
            handlerThread.join();
            handlerThread = null;
            handler = null;
        } catch (final InterruptedException e) {
            Log.e(TAG, "Exception!" + e.toString());
        }

        super.onPause();
        disableCamera();
    }

    @Override
    public synchronized void onResume() {
        super.onResume();
        if (!OpenCVLoader.initDebug()) {
            Log.d(
                TAG,
                "Internal OpenCV library not found. Using OpenCV Manager for initialization"
            );
            OpenCVLoader.initAsync(
                OpenCVLoader.OPENCV_VERSION_3_0_0, this, mBaseLoaderCallback
            );
        } else {
            Log.d(TAG, "OpenCV library found inside package. Using it!");
            mBaseLoaderCallback.onManagerConnected(LoaderCallbackInterface.SUCCESS);
        }

        handlerThread = new HandlerThread("inference");
        handlerThread.start();
        handler = new Handler(handlerThread.getLooper());
    }

    public void disableCamera() {
        if (mCameraView != null)
            mCameraView.disableView();
    }

    void startInference() {
        runInBackground(
            new Runnable() {
                @Override
                public void run() {
                Log.d(TAG, "Starting inference image processing");
                mProcessing = true;
                // Crop
                int side = 0;
                int x = 0;
                int y = 0;
                if (mPreviewHeight < mPreviewWidth) {
                    side = mPreviewHeight;
                    x = (mPreviewWidth - side) / 2;
                    y = 0;
                } else {
                    side = mPreviewWidth;
                    y = (mPreviewHeight - side) / 2;
                    x = 0;
                }
                Rect roi = new Rect(x, y, side, side);
                Mat img = new Mat(mRecentFrame, roi);

                // Resize
                Size size = new Size(INPUT_SIZE, INPUT_SIZE);
                Imgproc.resize(img, img, size);

                // Rotate
                int angle = mOrientation;
                if (angle >= 45 && angle < 135) {
                    // 90 cw
                    Core.flip(img.t(), img, 1);
                } else if (angle >= 135 && angle < 225) {
                    // 180
                    Core.flip(img, img, -1); 
                } else if (angle >= 225 && angle < 315) {
                    // 90 ccw
                    Core.flip(img.t(), img, 0);
                }

                // Convert
                Utils.matToBitmap(img, mInputBitmap);

                mProcessing = false;

                Log.d(TAG, "Starting inference");
                final long startTime = SystemClock.uptimeMillis();
                final List<Classifier.Recognition> results = mClassifier.recognizeImage(mInputBitmap);
                long lastProcessingTimeMs = SystemClock.uptimeMillis() - startTime;

                // Output resluts
                String output = VISUAL;
                for (Classifier.Recognition res : results) {
                    output += res.toString() + "\n";
                }
                output += lastProcessingTimeMs + " ms";

                setInference(output);
                startInference();
            }
        });
    }

    public void onCameraViewStarted(int width, int height) {
        // Calculate ratio to stretch camera preview on screen
        DisplayMetrics displayMetrics = new DisplayMetrics();
        getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        int screenHeight = displayMetrics.heightPixels;
        int screenWidth = displayMetrics.widthPixels;
        float ratio = 0;
        float heightRatio = 0;
        float widthRatio = 0;
        if (width < screenWidth)
            widthRatio = (float) screenWidth / (float) width;
        if (height < screenHeight)
            heightRatio = (float) screenHeight / (float) height;
        ratio = (heightRatio > widthRatio) ? heightRatio : widthRatio;
        mCameraView.setScale(ratio);

        mPreviewWidth = width;
        mPreviewHeight = height;

        // Create empty bitmaps until real camera frames arrived
        mInputBitmap = Bitmap.createBitmap(INPUT_SIZE, INPUT_SIZE, Config.ARGB_8888);
        Bitmap empty = Bitmap.createBitmap(mPreviewWidth, mPreviewHeight, Config.ARGB_8888);
        mRecentFrame = new Mat(mPreviewWidth, mPreviewHeight, CvType.CV_8UC4);
        Utils.bitmapToMat(empty, mRecentFrame);

        handlerThread = new HandlerThread("inference");
        handlerThread.start();
        handler = new Handler(handlerThread.getLooper());
        startInference();
    }

    public void onCameraViewStopped() {
    }

    public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {
        Mat image = inputFrame.rgba();
        if (!mProcessing)
            mRecentFrame = image.clone();
        salt(image.getNativeObjAddr());
        return image;
    }

    public void setInference(String letters) {
        final String inference = letters;
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                mVisuals.setResults(inference);
            }
        });
    }

    protected synchronized void runInBackground(final Runnable r) {
        if (handler != null) {
            handler.post(r);
        }
    }
    
    @Override
    protected synchronized void onDestroy() {
        super.onDestroy();
        disableCamera();
    }

    private static class ObbState {
        public StorageManager storageManager;
        public CharSequence status;
        public CharSequence path;
        ObbState(StorageManager storageManager, CharSequence status, CharSequence path) {
            this.storageManager = storageManager;
            this.status = status;
            this.path = path;
        }
    }

    public native void salt(long image);
    public native double setdir(String path);

}

File: app/src/main/java/uk/co/mishurov/termik2/MainActivity.kt
package uk.co.mishurov.termik2


import java.io.File
import java.io.BufferedInputStream
import java.io.FileOutputStream
import java.io.IOException

import android.Manifest
import android.os.Bundle
import android.os.Handler
import android.os.HandlerThread
import android.os.SystemClock
import android.support.v4.app.ActivityCompat
import android.view.WindowManager
import android.view.OrientationEventListener
import android.view.Surface
import android.view.View
import android.view.SurfaceView
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.ViewGroup
import android.hardware.SensorManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.PixelFormat
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.content.res.AssetManager
import android.preference.PreferenceManager
import android.util.DisplayMetrics

import com.google.vr.sdk.base.GvrActivity
import com.google.vr.sdk.base.GvrView

import org.opencv.android.BaseLoaderCallback
import org.opencv.android.CameraBridgeViewBase
import org.opencv.android.LoaderCallbackInterface
import org.opencv.android.OpenCVLoader
import org.opencv.android.Utils
import org.opencv.core.Mat
import org.opencv.core.Size
import org.opencv.core.Scalar
import org.opencv.core.Rect
import org.opencv.core.Core
import org.opencv.core.CvType
import org.opencv.imgproc.Imgproc

import org.opencv.dnn.Net
import org.opencv.dnn.Dnn

import org.tensorflow.classifier.Classifier
import org.tensorflow.classifier.TensorFlowImageClassifier

import android.util.Log


class MainActivity : GvrActivity(), View.OnTouchListener,
                            CameraBridgeViewBase.CvCameraViewListener2
{
    // Orientation
    private var mOrientationListener: OrientationEventListener? = null
    private var mListener: OnSharedPreferenceChangeListener? = null
    private var mPrefs: SharedPreferences? = null
    private var mEnginePref : Int = 0
    private var mOutputPref : Int = 0
    private var mOrientation = 0
    private var mScreenRotation = 0

    private var handler: Handler? = null
    private var handlerThread: HandlerThread? = null
    private var mPreviewWidth = 0
    private var mPreviewHeight = 0
    private var mScreenWidth = 0
    private var mScreenHeight = 0
    private var mProcessing = false
    private var mCameraView: KotlinCameraView? = null
    private var mVisuals: ResultsView? = null
    private var mGestureDetector: GestureDetector? = null

    private var mInputBitmap: Bitmap? = null
    private var mRecentFrame: Mat? = null
    // TensorFlow
    private var mClassifier: Classifier? = null
    // OpenCV DNN
    private var mNet: Net? = null
    private var mLabels: List<String>? = null
    private var mInputMat: Mat? = null

    // VR
    private var gvrRenderer: GvrRenderer? = null

    private val mBaseLoaderCallback = object : BaseLoaderCallback(this)
    {
        override fun onManagerConnected(status: Int)
        {
            when (status) {
                LoaderCallbackInterface.SUCCESS -> {
                    Log.i(TAG, "OpenCV loaded successfully")
                    // Load ndk built module, as specified in moduleName in build.gradle
                    // after opencv initialization
                    System.loadLibrary("processing")

                    if (mEnginePref == 0) {
                        setUpTensorFlow()
                    } else {
                        setUpOpenCvDnn()
                    }

                    mCameraView?.enableView()
                }
                else -> {
                    super.onManagerConnected(status)
                }
            }
        }
    }

    private fun getPath(file: String): String {

        val assetManager: AssetManager = getAssets()
        var inputStream: BufferedInputStream?

        try {
            val actual_file = file.split("file:///android_asset/")[1]
            inputStream = BufferedInputStream(assetManager.open(actual_file))

            val file_data: ByteArray = inputStream.readBytes()

            inputStream.read(file_data)
            inputStream.close()

            val outFile : File = File(getFilesDir(), actual_file)

            val os : FileOutputStream = FileOutputStream(outFile)
            os.write(file_data)
            os.close()

            return outFile.getAbsolutePath()

        } catch (e: IOException) {
            Log.i(TAG, "Failed to open NN file")
        }

        return ""
    }

    private fun readLabels(file: String) {
        val assetManager: AssetManager = getAssets()
        var inputStream: BufferedInputStream?

        try {
            val actual_file = file.split("file:///android_asset/")[1]
            inputStream = BufferedInputStream(assetManager.open(actual_file))
            val reader = inputStream.bufferedReader()
            mLabels = reader.readLines()

        } catch (e: IOException) {
            Log.i(TAG, "Failed to open labels file")
        }
    }

    private fun setUpOpenCvDnn() {
        mNet = Dnn.readNetFromTensorflow(getPath(MODEL_FILENAME))
        readLabels(LABEL_FILENAME)
    }


    private fun setUpTensorFlow() {
        mClassifier = TensorFlowImageClassifier.create(
                getAssets(),
                MODEL_FILENAME,
                LABEL_FILENAME,
                INPUT_SIZE,
                IMAGE_MEAN,
                IMAGE_STD,
                INPUT_NAME,
                OUTPUT_NAME
        )
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        window.addFlags(
                WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON
        )
        ActivityCompat.requestPermissions(
                this@MainActivity,
                arrayOf(Manifest.permission.CAMERA),
                1
        )

        mPrefs = PreferenceManager.getDefaultSharedPreferences(this)
        mListener = object : OnSharedPreferenceChangeListener {
            override fun onSharedPreferenceChanged(prefs: SharedPreferences,
                                                   key: String) {
                setPrefs(prefs)
            }
        }
        mPrefs?.registerOnSharedPreferenceChangeListener(mListener)
        PreferenceManager.setDefaultValues(this, R.xml.preferences, false)
        setPrefs(mPrefs!!)

        mGestureDetector = GestureDetector(this, GestureListener())

        val display = windowManager.defaultDisplay
        val rotation = display.rotation
        when (rotation) {
            Surface.ROTATION_90 -> mScreenRotation = 90
            Surface.ROTATION_180 -> mScreenRotation = 180
            Surface.ROTATION_270 -> mScreenRotation = -90
            else -> mScreenRotation = 0
        }

        // Listen orientation
        mVisuals = findViewById<ResultsView>(R.id.results)
        mVisuals?.setResults(VISUAL + "Calculating...")

        mCameraView = findViewById<KotlinCameraView>(R.id.camera)
        mCameraView?.setCvCameraViewListener(this)


        gvrView = findViewById<GvrView>(R.id.vr)
        if (mOutputPref == 2) {
            gvrView?.setDistortionCorrectionEnabled(false)
        }

        gvrView?.setEGLConfigChooser(8, 8, 8, 8, 0, 0)
        gvrRenderer = GvrRenderer(gvrView!!)
        setGvrView(gvrView!!)
        if (mOutputPref != 0) {
            gvrView?.setVisibility(SurfaceView.VISIBLE)
            gvrView?.setOnTouchListener(this)
        } else {
            gvrView.setStereoModeEnabled(false)
            gvrView?.setVisibility(SurfaceView.GONE)
        }

        val displayMetrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(displayMetrics)
        mScreenWidth = displayMetrics.widthPixels
        mScreenHeight = displayMetrics.heightPixels

        if (mOutputPref != 0) {
            val camParams = mCameraView!!.getLayoutParams()
            camParams.width = mScreenWidth
            camParams.height = mScreenHeight * 2
            mCameraView?.setLayoutParams(camParams)

            // make view trasparent since the image is rendered in GvrView
            mCameraView?.setZOrderOnTop(true)
            val holder = mCameraView?.getHolder()
            holder?.setFormat(PixelFormat.TRANSPARENT)

            val visParams = mVisuals!!.getLayoutParams()
            visParams.width = mScreenWidth / 2
            visParams.height = mScreenHeight
            mVisuals?.setLayoutParams(visParams)
            mVisuals?.setVrStyle()
        }

        mOrientationListener = object : OrientationEventListener(this,
                SensorManager.SENSOR_DELAY_NORMAL) {
            override fun onOrientationChanged(orientation: Int) {
                var orient = orientation
                orient += mScreenRotation
                this@MainActivity.mOrientation = orient
                this@MainActivity.mVisuals?.adjust(orient)
            }
        }

        if (mOrientationListener!!.canDetectOrientation()) {
            Log.v(TAG, "Can detect orientation")
            mOrientationListener?.enable()
        } else {
            Log.v(TAG, "Cannot detect orientation")
            mOrientationListener?.disable()
        }
    }

    override fun onPause() {
        gvrView!!.onPause()
        handlerThread!!.quitSafely()
        try {
            handlerThread!!.join()
            handlerThread = null
            handler = null
        } catch (e: InterruptedException) {
            Log.e(TAG, "Exception!" + e.toString())
        }
        super.onPause()
        disableCamera()
    }

    override fun onResume() {
        gvrView!!.onResume()
        super.onResume()
        if (!OpenCVLoader.initDebug()) {
            Log.d(
                    TAG,
                    "Internal OpenCV library not found. Using OpenCV Manager for initialization"
            )
            OpenCVLoader.initAsync(
                    OpenCVLoader.OPENCV_VERSION_3_0_0, this, mBaseLoaderCallback
            )
        } else {
            Log.d(TAG, "OpenCV library found inside package. Using it!")
            mBaseLoaderCallback.onManagerConnected(
                LoaderCallbackInterface.SUCCESS
            )
        }

        handlerThread = HandlerThread("inference")
        handlerThread?.start()
        handler = Handler(handlerThread?.looper)
    }

    fun disableCamera() {
        if (mCameraView != null)
            mCameraView?.disableView()
    }

    private fun openCvInference(): MutableList<String> {
        Imgproc.cvtColor(mInputMat, mInputMat, Imgproc.COLOR_RGBA2RGB)
        val size = Size(INPUT_SIZE.toDouble(), INPUT_SIZE.toDouble())
        val mean = Scalar(
            IMAGE_MEAN.toDouble(),
            IMAGE_MEAN.toDouble(),
            IMAGE_MEAN.toDouble()
        )
        val blob = Dnn.blobFromImage(mInputMat, 1.0, size, mean, true, true)

        mNet?.setInput(blob, INPUT_NAME)
        //val outputName = "softmax2"
        var result = mNet?.forward()
        result = result?.reshape(1, 1)

        var indices = result?.clone()
        Core.sortIdx(result, indices, Core.SORT_DESCENDING)

        var output: MutableList<String> = ArrayList()
        for (i in 0..4) {
            val index = indices?.get(0, i)?.get(0)!!.toInt()
            val label = mLabels!![index]
            val probability = result?.get(0, index)?.get(0)
            val probStr = "%.1f".format(probability?.times(100))
            output.add(label + " (" + probStr + "%)")
        }
        return output
    }

    private fun startInference() {
        runInBackground(
                Runnable {
                    Log.d(TAG, "Starting inference image processing")
                    mProcessing = true
                    // Crop
                    var side : Int
                    var x : Int
                    var y : Int
                    if (mPreviewHeight < mPreviewWidth) {
                        side = mPreviewHeight
                        x = (mPreviewWidth - side) / 2
                        y = 0
                    } else {
                        side = mPreviewWidth
                        y = (mPreviewHeight - side) / 2
                        x = 0
                    }
                    val roi = Rect(x, y, side, side)
                    mInputMat = Mat(mRecentFrame, roi)

                    // Resize
                    val size = Size(
                        INPUT_SIZE.toDouble(), INPUT_SIZE.toDouble()
                    )
                    Imgproc.resize(mInputMat!!, mInputMat, size)

                    // Rotate
                    val angle = mOrientation
                    if (angle >= 45 && angle < 135) {
                        // 90 cw
                        Core.flip(mInputMat!!.t(), mInputMat, 1)
                    } else if (angle >= 135 && angle < 225) {
                        // 180
                        Core.flip(mInputMat!!, mInputMat, -1)
                    } else if (angle >= 225 && angle < 315) {
                        // 90 ccw
                        Core.flip(mInputMat!!.t(), mInputMat, 0)
                    }

                    // Convert
                    Utils.matToBitmap(mInputMat!!, mInputBitmap)

                    mProcessing = false

                    Log.d(TAG, "Starting inference")
                    val startTime = SystemClock.uptimeMillis()

                    var output = VISUAL

                    if (mEnginePref == 0) {
                        val results = mClassifier?.recognizeImage(mInputBitmap)
                        for (res in results!!) {
                            output += res.toString().split("]")[1] + "\n"
                        }
                    } else {
                        val results = openCvInference()
                        for (res in results) output += res + "\n"
                    }

                    val lastProcessingTimeMs = SystemClock.uptimeMillis() - startTime

                    output += lastProcessingTimeMs.toString() + " ms"

                    setInference(output)
                    startInference()
                })
    }

    override fun onCameraViewStarted(width: Int, height: Int) {
        // Calculate ratio to stretch camera preview on screen
        var heightRatio = 0.0f
        var widthRatio = 0.0f

        if (width < mScreenWidth) {
            widthRatio = mScreenWidth.toFloat() / width.toFloat()
        }
        if (height < mScreenHeight) {
            heightRatio = mScreenHeight.toFloat() / height.toFloat()
        }

        val ratio = if (heightRatio > widthRatio) heightRatio else widthRatio
        mCameraView!!.setScale(ratio)

        mPreviewWidth = width
        mPreviewHeight = height

        mInputBitmap = Bitmap.createBitmap(
            INPUT_SIZE, INPUT_SIZE, Bitmap.Config.ARGB_8888
        )
        val empty = Bitmap.createBitmap(
            mPreviewWidth, mPreviewHeight, Bitmap.Config.ARGB_8888
        )
        mRecentFrame = Mat(mPreviewWidth, mPreviewHeight, CvType.CV_8UC4)
        Utils.bitmapToMat(empty, mRecentFrame)

        handlerThread = HandlerThread("inference")
        handlerThread?.start()
        handler = Handler(handlerThread?.looper)
        startInference()
    }

    override fun onCameraViewStopped() {}

    override fun onCameraFrame(
                    inputFrame: CameraBridgeViewBase.CvCameraViewFrame): Mat?
    {
        val image = inputFrame.rgba()
        if (!mProcessing)
            mRecentFrame = image.clone()
        process(image.nativeObjAddr)

        if (mOutputPref > 0) {
            var bmp = Bitmap.createBitmap(
                mPreviewWidth, mPreviewHeight, Bitmap.Config.ARGB_8888
            )
            Utils.matToBitmap(image, bmp)

            val ratio = mScreenHeight.toFloat() / mPreviewHeight.toFloat()

            bmp = Bitmap.createScaledBitmap(
               bmp,
               (mPreviewWidth * ratio).toInt() + 1,
               (mPreviewHeight * ratio).toInt() + 1,
               true
            )

            val offset = (mPreviewWidth * ratio - mScreenWidth * 0.5f) * 0.5f

            bmp = Bitmap.createBitmap(
               bmp, offset.toInt(), 0, mScreenWidth / 2, mScreenHeight
            )

            var vis = Bitmap.createBitmap(
                mScreenWidth / 2, mScreenHeight, Bitmap.Config.ARGB_8888
            )
            //vis.eraseColor(Color.argb(128, 255, 255, 255))
            val c = Canvas(vis)
            mVisuals?.drawText(c)

            gvrRenderer?.setBmp(bmp)
            gvrRenderer?.setVis(vis)
        }

        return if (mOutputPref > 0) null else image
    }

    fun setPrefs(prefs: SharedPreferences) {
        val engine : String = prefs.getString("engine_preference", "")
        mEnginePref = engine.toInt()
        val output : String = prefs.getString("output_preference", "")
        mOutputPref = output.toInt()
    }

    fun setInference(letters: String) {
        runOnUiThread { mVisuals?.setResults(letters) }
    }

    @Synchronized protected fun runInBackground(r: Runnable) {
        if (handler != null) {
            handler?.post(r)
        }
    }

    @Synchronized override fun onDestroy() {
        super.onDestroy()
        disableCamera()
    }

    override fun onTouchEvent(me: MotionEvent): Boolean
    {
        return mGestureDetector?.onTouchEvent(me) as Boolean
    }

    override fun onTouch(v: View, me: MotionEvent): Boolean
    {
        return mGestureDetector?.onTouchEvent(me) as Boolean
    }

    private inner class GestureListener
                            : GestureDetector.SimpleOnGestureListener() {
        override fun onDoubleTap(e: MotionEvent): Boolean {
            val intent = Intent(
                    this@MainActivity,
                    SettingsActivity::class.java
            )
            startActivity(intent)
            return true
        }
    }

/*
    override fun onBackPressed() {
    }

    override fun onCardboardTrigger() {
    }
*/
    external fun process(image: Long)

    companion object {

        private val TAG = "Termik"

        private val VISUAL = "Visual:\n"
        private val MODEL_FILENAME = "file:///android_asset/tensorflow_inception_graph.pb"
        private val LABEL_FILENAME = "file:///android_asset/imagenet_comp_graph_label_strings.txt"
        private val INPUT_SIZE = 224
        private val IMAGE_MEAN = 117
        private val IMAGE_STD = 1f
        private val INPUT_NAME = "input"
        private val OUTPUT_NAME = "output"
    }

}


